### PR TITLE
Remove core restriction in cache and turn it into an active/standby restriction instead

### PR DESCRIPTION
**Pull request by jefferai**

### CODE REVIEW

File: vault/core.go [hunk 1]
```diff
@@ -1573,10 +1575,8 @@ func (c *Core) postUnseal() (retErr error) {
 	c.clearForwardingClients()
 	c.requestForwardingConnectionLock.Unlock()
 
-	// Purge the backend if supported
-	if purgable, ok := c.physical.(physical.Purgable); ok {
-		purgable.Purge(c.activeContext)
-	}
+	c.physicalCache.Purge(c.activeContext)
+	c.physicalCache.SetEnabled(true)
```
Comment (briankassouf) [inline vault/core.go]: I think we want to conditionally enable it based on if we have caching disabled in the config

Comment (jefferai) [reply vault/core.go]: Oops yes you're right. I thought about that before and then forgot to do it.

Comment (jefferai) [reply vault/core.go]: Fixed!

File: physical/physical.go [hunk 2]
```diff
@@ -56,10 +56,11 @@ type HABackend interface {
 	HAEnabled() bool
 }
 
-// Purgable is an optional interface for backends that support
-// purging of their caches.
-type Purgable interface {
+// An interface for backends that can toggle on or off special functionality
```
Comment (calvn) [inline physical/physical.go]: Small nit, but it should start with "ToggleablePurgemonster is ..."

File: physical/inmem/cache_test.go [hunk 3]
```diff
@@ -63,268 +64,3 @@ func TestCache_Purge(t *testing.T) {
 		t.Fatalf("should not have key")
 	}
 }
-
```
Comment (calvn) [inline physical/inmem/cache_test.go]: Should we add test cases for when cache.SetEnabled(false) to make sure operations are not getting/modifying from cached values?

Comment (jefferai) [reply physical/inmem/cache_test.go]: I can add such a test, good idea.

File: physical/cache.go [hunk 4]
```diff
@@ -34,49 +32,43 @@ type TransactionalCache struct {
...
 	c := &Cache{
-		backend:    b,
-		lru:        cache,
-		locks:      locksutil.CreateLocks(),
-		exceptions: cacheExceptions,
-		logger:     logger,
+		backend: b,
+		lru:     cache,
+		locks:   locksutil.CreateLocks(),
+		logger:  logger,
```
Comment (calvn) [inline physical/cache.go]: `enabled` seems to be false by default (meaning no cache), is this intentional?

Comment (jefferai) [reply physical/cache.go]: Yep. Fail safe.

Comment (calvn) [reply physical/cache.go]: Ahh, in that case perhaps adding a comment on the need to call SetEnabled before the Cache really starts caching would be helpful.

Comment (jefferai) [reply physical/cache.go]: Done

File: physical/inmem/cache_test.go (hunk 3 above)
Comment (jefferai) [reply physical/inmem/cache_test.go]: Done!

File: physical/physical.go [hunk 5]
```diff
@@ -56,10 +56,11 @@ type HABackend interface {
 	HAEnabled() bool
 }
 
-// Purgable is an optional interface for backends that support
-// purging of their caches.
-type Purgable interface {
+// ToggleablePurgemonsteris an interface for backends that can toggle on or off
```
Comment (chrishoffman) [inline physical/physical.go]: s/ToggleablePurgemonsteris/ToggleablePurgemonster is/

**Pull request merged**